fix(parser): Ban long flags with literals

rami3l opened the pull request:
Closes #1543.

The objective of this PR is to ban use of literals with long flags (long args without `TakesValue`), eg. `--flag=literal`, in order to avoid the current confusion that `--flag=false` is considered as `true`.

Ideally, `--flag=literal` could be re-enabled for certain values of `literal`, but the choice of such values is left for further discussion (in #1649 for CLI flags, and #2539 for environment variables).

### Unresolved Questions

- [x] ~~How should this interact with multiple occurrences?~~

epage commented:
>  How should this interact with multiple occurrences?

I wish overrides-self was a default :).  It'd especially make sense in this case but people can still relatively easily opt-in, if they know how.

epage commented:
Looking over #1649, it doesn't look like there was consensus on moving forward with this.  It also looked like it was possible to implement this by users in clap today.  

I'm particularly concerned about
- User confusion with an argument that can only be used as `--arg=value` and not also `-avalue` or `--arg value`
- The needed special casing of the parser for this
- If we did support this as a `--arg value`, it'd then need to be opt-in with a flag because conditional number of values can have surprising results.

rami3l commented:
@epage As a clap user I truly think that the solution proposed in https://github.com/clap-rs/clap/issues/1649#issuecomment-661274943 is acceptable. However, the problem that I really want to solve here is that `--flag=false` is considered as `true`, as good library should always minimize its possibility to be misused.

I have an idea: to avoid possible argues about `true/false` against `yes/no` and `1/0`, maybe we can do this in another way by banning `--flag=something` altogether, and somehow guide our users towards the `pseudo-flag` pattern, also to give our users more control?

epage commented:
> However, the problem that I really want to solve here is that --flag=false is considered as true

... I didn't want to believe this but I threw in a test real quick
```rust
    assert_eq!(
        Opt { alice: true },
        Opt::parse_from(&["test", "--alice=false"])
    );
```

imo thats bad and seems like a bug in the parser.  If there is a `val`, we shouldn't ignore it but should error if we are processing a flag.

rami3l commented:
@pksunkara After some consideration, at this point I'm leaning towards modifying my original code to ban `--flag=something` altogether. Is this desirable?

rami3l commented:
@epage @pksunkara  So I have just banned all usages of flags with a literal, giving a `too_many_values` error.

The rest of this problem is to somehow make users of this crate more aware of the `pseudo-flag` approach. Any advice?

If this gets accepted, I will then make another PR on env variables.

epage commented:
> The rest of this problem is to somehow make users of this crate more aware of the pseudo-flag approach. Any advice?

imo we should have a place to document common arg patterns.  Unsure if `examples/` is the best place or the website.

ldm0 commented on src/parse/parser.rs:
```diff
@@ -1061,11 +1061,39 @@ impl<'help, 'app> Parser<'help, 'app> {
-                return Ok(self.parse_flag(opt, matcher));
             }
+            debug!("Parser::parse_long_arg: Found a flag");
+            self.check_for_help_and_version_str(&arg)?;
```
Suggestion: I think clap should emits too many values error in this situation: `--help=xxx`, which means `check_for_help_and_version_str` should be placed in the `None` branch.

rami3l replied:
Oops! I'm sorry, I seem to have forgotten the `AS::InferLongArgs` part. I'll fix that too.

rami3l replied:
Resolved.

epage commented:
Could you update your PR to say its resolving #1543 rather than #1649

rami3l commented:
@epage I updated the description mentioning both issues: there should be no need to make a choice between them.

@ldm0 I modified one of the existing examples hoping to make more users aware of the `pseudo-flag` pattern.

epage commented:
>  I updated the description mentioning both issues: there should be no need to make a choice between them

I take it you are proposing we close #1649 out by rejecting the proposed idea and instead document how to use existing clap APIs?

As I said earlier, I hadn't seen consensus on it.  No idea what clap's decision making process is like to say whether documenting is a way of closing #1649 or to just unblock people until a decision is made.  This is why I was assuming #1649 was going to be taken out of this PR.  

One thing gained by decoupling the decision over #1649 from this PR is it means it doesn't have to be bogged down before merging useful improvements.

rami3l commented:
@epage Thanks for clarifying.

rami3l commented:
@epage Sorry, clearly I wasn't very well thought out. It turns out that soon we might be discussing the very same problem over again with env variables...

pksunkara commented on examples/05_flag_args.rs:
```diff
@@ -29,6 +29,15 @@ fn main() {
+                .long("pseudo-flag")
+                .min_values(0)
+                .max_values(1),
+        )
```
Let's remove this chunk.

rami3l replied:
Done.

pksunkara commented on examples/05_flag_args.rs:
```diff
@@ -38,6 +47,11 @@ fn main() {
+    if matches.value_of("pseudo-flag") != Some("false") {
+        println!("Pseudo-flag is turned on");
+    }
+
```
And this one.

rami3l replied:
Could you please explain why? I copied this directly from https://github.com/clap-rs/clap/issues/1649#issuecomment-661274943.

pksunkara replied:
Because this PR is not related to #1649 at all.

rami3l replied:
I'm aware of the fact that this PR is unrelated, although I'm following the advice presented in https://github.com/clap-rs/clap/pull/2619#issuecomment-888433774:

> > The rest of this problem is to somehow make users of this crate more aware of the pseudo-flag approach. Any advice?
> 
> imo we should have a place to document common arg patterns. Unsure if `examples/` is the best place or the website.

pksunkara replied:
I am not sure why that was even discussed in the first place. Let's just remove it for now and add it to examples when actually resolving #1649

rami3l replied:
Reverted.

pksunkara commented on src/build/app/mod.rs:
```diff
@@ -2485,8 +2485,7 @@ impl<'help> App<'help> {
             self.subcommands.push(
-                App::new("help")
-                    .about("Print this message or the help of the given subcommand(s)"),
+                App::new("help").about("Print this message or the help of the given subcommand(s)"),
```
This is fixed in master.

rami3l replied:
Reverted.

pksunkara commented on tests/flags.rs:
```diff
@@ -87,6 +87,24 @@ fn flag_using_long() {
     assert!(m.is_present("color"));
 }
 
+#[cfg(feature = "suggestions")]
```
Why is feature required?

rami3l replied:
Removed.

pksunkara commented on tests/flags.rs:
```diff
@@ -87,6 +87,24 @@ fn flag_using_long() {
+    use clap::ErrorKind;
+
+    let m = App::new("flag")
+        .args(&[Arg::from("--rainbow 'yet another flag'").multiple_occurrences(true)])
```
Can you please use `Arg::new` instead of `Arg::from` and `arg` instead of `args`? And let's remove `multiple_occurrences`.

rami3l replied:
Done.

pksunkara commented on tests/flags.rs:
```diff
@@ -87,6 +87,24 @@ fn flag_using_long() {
+            "--rainbow",
+            "--rainbow",
+            "--rainbow=false",
+            "--rainbow=true",
```
Update this to just `vec!["--rainbow=false"]`.

rami3l replied:
I guess you meant `vec!["", "--rainbow=false"]`.

pksunkara commented on tests/flags.rs:
```diff
@@ -87,6 +87,17 @@ fn flag_using_long() {
+    use clap::ErrorKind;
+
+    let m = App::new("flag")
+        .arg(Arg::new("--rainbow 'yet another flag'"))
```
When I meant use `Arg::new`, I meant use the arg builder methods, not use the usage strings to build the arg.

rami3l replied:
Sorry, now it's fixed.

pksunkara reviewed (changes requested):
The example file still has changes.
